### PR TITLE
🎉 New BigQuery destination with Structured/Repeated Records

### LINKED ISSUE #1927: Load JSON data into BigQuery as structured data (STRUCT type)
## Tell us about the problem you're trying to solve
I'd like to the structure of the JSON in my data source (e.g. Google AdWords) to be available in my destination, BigQuery, so that I can run meaningful queries on the data.  However, with Airbyte, I only have two options:
(a) Load the JSON data as a big string in one `_airbyte_data` column, which I can't query
(b) Use Basic Normalization, which doesn't handle hierarchical data

## Describe the solution you’d like
BigQuery is already built to handle hierarchical data, so I'd love to have JSON simply load into BigQuery as structured records.  The usual `bq` command can do this, for example: https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-json#loading_json_data_into_a_new_table

Note that it's necessary to provide a schema.  I would recommend using https://pypi.org/project/bigquery-schema-generator/, which will scan the file and produce a schema.  It's worked great for me so far.  The `–autodetect` option to `bq` won't cut it, because it only scans 100 rows, so if a new field appears in some other row, the entire import will fail.  `bigquery-schema-generator` will scan the whole file and come up with a schema that can accommodate every record.

## Describe the alternative you’ve considered or used
The way I've usually done this is to export the data from my data source into a big JSONL file (newline-delimited JSON), then run `bigquery-schema-generator` over it to produce a JSON schema file.  Then I use the BigQuery console to load the data; it takes JSONL as one of its source formats and accepts a JSON schema, and that works nicely.

In theory, I could work around the problem by exporting all the data from the `_airbyte_data` column and then going through the same procedure of generating a schema and then manually uploading the data.  But it would make a lot more sense for Airbyte to simply load the data in its true, raw form.  Arguably, the "most raw" form is the JSON, not a table of rows that contain JSON serialized to strings.

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
Closes #1927 

## How
A new destination that does not rely on base-normalization but implements its own native normalization with BigQuery by converting JSON Schema into google cloud schema, and thus, handle structured//repeated records/arrays.

### Implementation notes

This issue was not doable from base-normalization python/dbt codebase as it seemed difficult (not possible?) to implement in BQ Standard SQL logic that parses a JSON column string into separate columns while building or casting to a STRUCT field... (the logic with nested is making this harder to think about too, and certainly not very efficient!)

It seemed more standard (from [google docs](https://cloud.google.com/bigquery/docs/nested-repeated)) to provide the JSON Schema to BigQuery at loading time instead when creating the tables. Therefore, do the implementation in the java codebase.

I started implementing this following the same pattern of CopyDestination that is first uploading to a cloud storage and then load the warehouse or directly to the warehouse depending on some configs values.

However, we then run into challenges to solve where conflicts can arise:
1. Run a BigQuery destination with struct/repeated fields
2. The destination could additionally persist into _airbyte_raw tables with JSON Blobs
3. Run normalization that will override what was produced by step 1 using what is produced by step 2 is counterproductive

Normalization code should probably be tweaked to be more easily compatible with the tables produced by this new denormalized destination instead. But that would mean more development work into the scope of this issue.

As a result, it is more straightforward to separate them into two distinct connectors for the moment with the "de-normalized" destination not being able to support normalization (and not supporting append-dedup either). In those cases, users can fallback to standard `destination-bigquery` or implement their own custom transformations post-sync instead for the moment.

When/if normalization is refactored to be compatible with de-normalized tables, then we could merge the two bigquery destinations back together into a single connector.

## Recommended reading order
1. `airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java`
2. `airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedRecordConsumer.java`
3.  `airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/JsonSchemaType.java`
4. `airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java`
5. the rest

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [x] Issue acceptance criteria met
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Secrets are annotated with `airbyte_secret` in output spec
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [x] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Documentation updated 
    - [ ] README
    - [ ] CHANGELOG.md
    - [ ] Reference docs in the `docs/integrations/` directory.
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [x] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationAcceptanceTest.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    }
+    return name.startsWith("_airbyte") && name.endsWith("_hashid");
+  }
+
+  private Object getTypedFieldValue(FieldValueList row, Field field) {
+    FieldValue fieldValue = row.get(field.getName());
+    if (fieldValue.getValue() != null) {
+      return switch (field.getType().getStandardType()) {
+        case FLOAT64, NUMERIC -> fieldValue.getDoubleValue();
+        case INT64 -> fieldValue.getNumericValue().intValue();
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationAcceptanceTest.java:184]: Same here. `INT64` should be mapped to `longValue()` I think.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+  @Override
+  protected Schema getBigQuerySchema(JsonNode jsonSchema) {
+    final List<Field> fieldList = getSchemaFields(getNamingResolver(), jsonSchema);
+    if (fieldList.stream().noneMatch(f -> f.getName().equals(JavaBaseConstants.COLUMN_NAME_AB_ID))) {
+      fieldList.add(Field.of(JavaBaseConstants.COLUMN_NAME_AB_ID, StandardSQLTypeName.STRING));
+    }
+    if (fieldList.stream().noneMatch(f -> f.getName().equals(JavaBaseConstants.COLUMN_NAME_EMITTED_AT))) {
+      fieldList.add(Field.of(JavaBaseConstants.COLUMN_NAME_EMITTED_AT, StandardSQLTypeName.TIMESTAMP));
+    }
```
Comment (tuliren) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:86]: Are the above two `if` checks always false? It seems that the original Json schema will never have the two Airbyte columns.

**Review by tuliren — COMMENTED** [review]
Looks good! Left a few minor comments.

Also this destination should have a doc.

File: airbyte-config/init/src/main/resources/config/STANDARD_DESTINATION_DEFINITION/079d5540-f236-4294-ba7c-ade8fd918496.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "destinationDefinitionId": "079d5540-f236-4294-ba7c-ade8fd918496",
+  "name": "BigQuery de-normalized",
```
Comment (cgardens) [inline airbyte-config/init/src/main/resources/config/STANDARD_DESTINATION_DEFINITION/079d5540-f236-4294-ba7c-ade8fd918496.json:3]:
```suggestion
  "name": "BigQuery (Typed Struct)",
```
I think this display name is a little bit clearer? Fine with me if you want to stick with denormalized but if you do it should be one word and I'd suggest putting it in parens. so `BigQuery (Denormalized)`.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/StandardNameTransformer.java [hunk 4]
```diff
@@ -41,7 +41,7 @@ public String getRawTableName(String streamName) {
 
   @Override
   public String getTmpTableName(String streamName) {
-    return convertStreamName("_airbyte_" + Instant.now().toEpochMilli() + "_" + getRawTableName(streamName));
+    return convertStreamName(Strings.addRandomSuffix("_airbyte_tmp", "_", 3) + "_" + streamName);
```
Comment (cgardens) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/StandardNameTransformer.java:44]: why is this needed?

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py [hunk 5]
```diff
@@ -389,9 +389,9 @@ def cast_property_type(self, property_name: str, column_name: str, jinja_column:
             print(f"WARN: Unknown type for column {property_name} at {self.current_json_path()}")
             return column_name
         elif is_array(definition["type"]):
-            return self.cast_property_type_as_array(property_name, column_name)
```
Comment (cgardens) [inline airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:392]: why is this changing? isn't this affecting the original BQ destination?

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BigQueryDenormalizedDestination extends BigQueryDestination {
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:55]: Could you explain some more why we wouldn't want to make this a `CopyDestination`? I understand why this is split from `BigQueryDestination`--that totally makes sense to me. But it seems like this should be able to support `CopyDestination` and normal insert?

**Review by cgardens — APPROVED** [review]
Looks good. I feel like this should support copy destination functionality too. If that's too hard to do now then let's at least create an issue to do it.

**Comment by marcosmarxm** [comment]
@ChristopheDuong i'd bumped a new version of the BigQuery destination last week, please merge master into your branch.

File: airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py (hunk 5 above)
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py:392]: This is doing the exact same thing as before (minus the extra function call) so it's not changing anything.

The extra function call was a placeholder to implement the struct/repeated "casting" there but it's actually not doable so not useful anymore

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java (hunk 2 above)
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:86]: the original JSON Schema can have the two airbyte columns if the streams were produced by airbyte and re-used as source streams

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/StandardNameTransformer.java (hunk 4 above)
Comment (ChristopheDuong) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/StandardNameTransformer.java:44]: it was originally needed when I was producing both a raw table (json blob) and struct typed table by the same destination, so I needed two different "tmp" names. But then I moved away from that approach... 

Should I revert to the old naming with timestamp?

File: airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java [hunk 7]
```diff
@@ -171,7 +171,7 @@ private Object getTypedFieldValue(FieldValueList row, Field field) {
     if (fieldValue.getValue() != null) {
       return switch (field.getType().getStandardType()) {
         case FLOAT64, NUMERIC -> fieldValue.getDoubleValue();
-        case INT64 -> fieldValue.getLongValue();
+        case INT64 -> fieldValue.getNumericValue().intValue();
```
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java:174]: Tests are failing when it's used with `fieldValue.getLongValue();`

```
expected: com.fasterxml.jackson.databind.node.IntNode@4843dc7<1> but was: com.fasterxml.jackson.databind.node.LongNode@f72b399b<1>
org.opentest4j.AssertionFailedError: expected: com.fasterxml.jackson.databind.node.IntNode@4843dc7<1> but was: com.fasterxml.jackson.databind.node.LongNode@f72b399b<1>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1124)
	at io.airbyte.integrations.standardtest.destination.DestinationAcceptanceTest.assertSameData(DestinationAcceptanceTest.java:985)
	at io.airbyte.integrations.standardtest.destination.DestinationAcceptanceTest.assertSameMessages(DestinationAcceptanceTest.java:955)
	at io.airbyte.integrations.standardtest.destination.DestinationAcceptanceTest.retrieveRawRecordsAndAssertSameMessages(DestinationAcceptanceTest.java:937)
	at io.airbyte.integrations.standardtest.destination.DestinationAcceptanceTest.testIncrementalSync(DestinationAcceptanceTest.java:475)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$178/000000008A80D220.apply(Unknown Source)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall$$Lambda$179/000000008A9AD420.apply(Unknown Source)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$$Lambda$315/00000000AB169E20.apply(Unknown Source)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$420/00000000ACB54220.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$214/0000000000000000.accept(Unknown Source)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$214/0000000000000000.accept(Unknown Source)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher$$Lambda$173/0000000000000000.accept(Unknown Source)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy4.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:851)
```

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java (hunk 6 above)
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:55]: I am not saying it shouldn't be a `CopyDestination`, I just tried a similar approach to `CopyDestination`, and I faced issues that did not make it easy to pursue for this use case.

`BigQueryDestination` isn't currently implemented for `CopyDestination`, and it was not the goal of this PR either.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedRecordConsumer.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  protected JsonNode formatData(FieldList fields, JsonNode root) {
+    if (root.isObject()) {
+      final List<String> fieldNames = fields.stream().map(Field::getName).collect(Collectors.toList());
+      return Jsons.jsonNode(Jsons.keys(root).stream()
+          .filter(key -> {
+            final boolean validKey = fieldNames.contains(namingResolver.getIdentifier(key));
+            if (!validKey) {
+              LOGGER.warn("Ignoring field {} as it is not defined in catalog", key);
```
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedRecordConsumer.java:92]: good catch, thanks

File: airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tuliren) [reply airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java:174]: > expected: com.fasterxml.jackson.databind.node.IntNode@4843dc7<1> but was: com.fasterxml.jackson.databind.node.LongNode@f72b399b<1>

I think this is because the unit test does not know that it should compare the value wrapped inside the node. But the values are the same. To bypass it, you can perform a deliberate conversion of `LongNode` to `IntNode` in the `retrieveRecords`.

But if we use integer value here, there will be number overflows in production.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java (hunk 2 above)
Comment (tuliren) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:86]: Oh, I see. Good to know this.

File: airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChristopheDuong) [reply airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java:174]: this piece of code is part of the test, not part of the connector, so this should have no effect on production?

[the remark this answers is not in the document]
Comment (tuliren) [reply airbyte-integrations/connectors/destination-bigquery/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDestinationAcceptanceTest.java:174]: Oh, sorry, I missed that it is in test. Should have no problem then.

**Review by subodh1810 — APPROVED** [review]

**Comment by ChristopheDuong** [comment]
/publish connector=connectors/destination-bigquery-denormalized
